### DISCUSSION TITLE
Hearbeat timeout and non-async Mojo action

### DISCUSSION THREAD

**Discussion by evengar2008** [mojolicious/mojo#1815] [Q&A]
The documentation
https://docs.mojolicious.org/Mojolicious/Guides/FAQ#What-does-Worker-31842-has-no-heartbeat-50-seconds-restarting-mean

says that "a blocking operation such as an infinite loop in your application can prevent this, and will force the affected worker to be restarted after a timeout."

Is it correct that if Mojo worker is executing some action with non-async Perl code which is taking longer than `heartbeat_timeout` to complete the worker will be eventually sent SIGQUIT, and, later on, killed with SIGKILL signal? 
And Mojo worker cannot send heartbeat to the master process until it finishes non-async action execution?

**Comment by kraih** [comment]
Why would the docs lie?

**Reply by vsespb** [reply]
But that means that synchronious code application should take additional care to work properly.
There is no "heartbeat" (in usual sense) in sync code.
And options graceful_timeout,heartbeat_interval,heartbeat_timeout should be setup in some tricky way.
For example If I want one request timeout to be 90 seconds, I should set up heartbeat_interval=90s,
heartbeat_timeout=1s,graceful_timeout=1s? Or.. I am not sure?
Is not this is something that should be documented? Or at least the fact that it works out of the box for async code only?

**Reply by kraih** [reply]
Sorry, there seems to a language barrier here, i don't understand what you mean.

**Comment by jberger** [comment]
> But that means that synchronious code application should take additional care to work properly.
There is no "heartbeat" (in usual sense) in sync code.
And options graceful_timeout,heartbeat_interval,heartbeat_timeout should be setup in some tricky way.
For example If I want one request timeout to be 90 seconds, I should set up heartbeat_interval=90s,
heartbeat_timeout=1s,graceful_timeout=1s? Or.. I am not sure?
Is not this is something that should be documented? Or at least the fact that it works out of the box for async code only?

In Mojolicious, web applications are necessarily asynchronous by nature. The fact that you can run blocking code on them is a natural result of how Perl operates, namely that there aren't non-blocking versions of each function. As a result, in your web application, even if you write synchronous code, you must be sure to allow your application the time it needs to heartbeat. If that means unblocking something long-running by using subprocess, using a timer rather than sleep, or increasing the heartbeat timeout to accomodate, that is up to you to do.

**Reply by vsespb** [reply]
Thanks! I still can suggest that there should be some doc for people using Mojo with synchronous codebases telling exactly what you told.
